test/configurationTests.ts

```diff
@@ -251,6 +251,26 @@ describe("Configuration", () => {
             const actualConfig = extendConfigurationFile(baseConfig, extendingConfig);
             assertConfigEquals(actualConfig, expectedConfig);
         });
+
+        it("overrides defaultSeverity of base configs", () => {
+            const config = loadConfigurationFromPath("./test/config/tslint-extends-default-severity-only-in-extended.json");
+            assert.equal<RuleSeverity | undefined>(
+                "warning",
+                config.rules.get("default-severity-unspecified")!.ruleSeverity,
+                "should apply defaultSeverity to base config with no defaultSeverity");
+            assert.equal<RuleSeverity | undefined>(
+                "warning",
+                config.rules.get("default-severity-error")!.ruleSeverity,
+                "should override defaultSeverity defined in base config");
+            assert.equal<RuleSeverity | undefined>(
+                "warning",
+                config.rules.get("default-severity-warning")!.ruleSeverity,
+                "should apply defaultSeverity to extending config");
+            assert.equal<RuleSeverity | undefined>(
+                "error",
+                config.rules.get("default-severity-only-in-extended")!.ruleSeverity,
+                "should not inherit defaultSeverity from base configs");
```

Comment:
I don't think this is quite right. Let's say I inherit from a shared config file where defaultSeverity is set to `"warning"`.

Then I change one of the rule configs to enable some specific rule option.

```json
{
  "extends": "config-where-failures-are-warnings",
  "rules": {
    "comment-format": {
      "options": ["check-space", "check-lowercase"]
    }
  }
}
```

now, comment-format produces errors. this is mildly confusing UX. While I think people could generally live with it, I think defaultSeverity should ideally be inherited from the `extendingConfig` if it is not specified in the current config.

Reply:
That's actually a bit more complicated to achive. We have a circular dependency here: we need to call `parseConfigFile` to know extended configs, but `parseConfigFile` would already need to know the `defaultSeverity` of the extended config file.
We need a bigger refactor of `configuration.ts` to make it work. I think this would result in a breaking change for API users.

Btw. which `defaultSeverity` is chosen when extending two or more config files? The last one?

Reply:
> which defaultSeverity is chosen when extending two or more config files? The last one?

yes, the last one in the `extends` list.

Reply:
Do we agree that this change is better to be done in a separate PR?

Reply:
@M0ns1gn0r I agree that this may be too much for a first-time contributor. Though the necessary changes may not be too big.
We should definitely add the missing bits shortly after landing this PR and before cutting a release.

@adidahiya we could make it work as you described by moving the extending of the config file from `loadConfigurationFromPath` into `parseConfigFile`. That way we could know the `defaultSeverity` of the parent config before it is used in `parseRuleOptions`.
IMO the best way to implement this backward compatible is by adding an optional parameter to `parseConfigFile`. That parameter is a function to load a config file by path. If this parameter is passed, `parseConfigFile` resolves all extended configs and reduces them. Otherwise the config is returned without touching `extends`.

Reply:
@ajafff your proposal sounds reasonable to me. We might want to do this work on a feature branch so that `master` is not left in an in-between state (the Github UI supports re-targeting PR branch destinations).